resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala

```diff
@@ -92,7 +101,9 @@ private[spark] object SparkKubernetesClientFactory extends Logging {
       .withRequestTimeout(clientType.requestTimeout(sparkConf))
       .withConnectionTimeout(clientType.connectionTimeout(sparkConf))
       .withTrustCerts(sparkConf.get(KUBERNETES_TRUST_CERTIFICATES))
-      .withOption(oauthTokenValue) {
+      .withOption(oauthTokenProviderInstance) {
+        (provider, configBuilder) => configBuilder.withOauthTokenProvider(provider)
+      }.withOption(oauthTokenValue) {
```

Comment:
To be clear, if we have two configurations at the same time mistakingly, we invoke `OAuthTokenProvider` and override it by `oauthTokenValue`?

Reply:
Could you check what happens when we call `configBuilder.withOauthTokenProvider(provider)` and `configBuilder.withOauthToken(token)` together?

Reply:
The token that was specified in `configBuilder.withOauthToken(token)` will stay unused, and the class that implements `OAuthTokenProvider` will be used instead. I confirmed this via tests on my cluster by simultaneously:

1 . Adding a debugging log into my class and providing it via 
```
--conf spark.kubernetes.client.oauth.token.provider.class <myClass>
```
2. Specifying tokens via 
```
--conf spark.kubernetes.authenticate.submission.oauthToken=<token>
--conf spark.kubernetes.authenticate.driver.oauthToken=<token> 
--conf spark.kubernetes.authenticate.oauthToken=<token>
```

My log:
```
[2021-08-11 12:51:58,012] DEBUG (OAuthGoogleTokenProvider.java:62) - Refreshing kubernetes oauth token via [/usr/lib/google-cloud-sdk/bin/gcloud, config, config-helper, --format=json]
[2021-08-11 12:51:58,498] DEBUG (OAuthGoogleTokenProvider.java:73) - New token expiry time is 3511s
```

From the log it could be concluded that `OAuthTokenProvider` has a higher precedence over the token that was specified in `configBuilder.withOauthToken(token) `.  It could also be confirmed by looking at the fabric's code in 
https://github.com/fabric8io/kubernetes-client/blob/74cc63df9b6333d083ee24a6ff3455eaad0a6da8/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/RequestConfig.java#L136

Reply:
Thank you for checking. In that case, the current situation may cause a confusion to the Spark users. From Spark perspective, we have two options.

- Option A: We can revise the description of this new configuration about that precedence.
- Option B: We can ignore `spark.kubernetes.client.oauth.token.provider.class` when `oauthToken` exists from Spark side.

Which one do you prefer?

Reply:
> the current situation may cause a confusion to the Spark users.

Indeed. Thanks for making me realize this. The patch is missing a way for a Spark user to specify how they would like the Spark to authenticate in Kubernetes when running on a client or cluster mode. There is several options present for oauthToken: 
```
spark.kubernetes.authenticate.submission.oauthToken
spark.kubernetes.authenticate.driver.oauthToken 
spark.kubernetes.authenticate.oauthToken
```
And I think we need to have the same for token provider, like `spark.kubernetes.authenticate.*.oauthTokenProvider`. 

This change would: 

- Make the new options consistent with the rest of kubernetes options.
- Make existing options mutually exclusive. For every mode only one of the `oauthToken`, `oauthTokenFile`, `oauthTokenProvider` would be allowed. 

If we try this, we won't have to add anything about precedence into the docs and there would be no need to ignore anything in the code. 

@dongjoon-hyun, sorry for a lot of text.

Reply:
I think either precedence or doing an assertion that only one of these is set is fine, but let's just pick one and do it so we can get this in.

Reply:
I prepared the changes, so now we have:

• `OAuthTokenProvider` options are now consistent with the other two options – `OAuthToken` and `OAuthTokenFile`.
• Added an assertion that only one of the options is set (either `OAuthToken`, `OAuthTokenFile` or `OAuthTokenProvider`
• Added docs

I have tested this in client mode on Kubernetes.